homeassistant/components/here_travel_time/coordinator.py

```diff
@@ -0,0 +1,302 @@
+            }
+        )
+
+
+def prepare_parameters(
+    hass: HomeAssistant,
+    config: HERETravelTimeConfig,
+) -> tuple[list[str], list[str], str | None, str | None]:
+    """Prepare parameters for the HERE api."""
+
+    def _from_entity_id(entity_id: str) -> list[str]:
+        coordinates = find_coordinates(hass, entity_id)
+        if coordinates is None:
+            raise InvalidCoordinatesException(f"No coordinates found for {entity_id}")
+        try:
+            formatted_coordinates = coordinates.split(",")
+            vol.Schema(cv.gps(formatted_coordinates))
+        except (AttributeError, vol.ExactSequenceInvalid) as ex:
+            raise InvalidCoordinatesException(
+                f"{coordinates} are not valid coordinates"
+            ) from ex
+        return formatted_coordinates
+
+    # Destination
+    if config.destination_entity_id is not None:
+        destination = _from_entity_id(config.destination_entity_id)
+    else:
+        destination = [
+            str(config.destination_latitude),
+            str(config.destination_longitude),
+        ]
+
+    # Origin
+    if config.origin_entity_id is not None:
+        origin = _from_entity_id(config.origin_entity_id)
+    else:
+        origin = [
+            str(config.origin_latitude),
+            str(config.origin_longitude),
+        ]
+
+    # Arrival/Departure
+    arrival: str | None = None
+    departure: str | None = None
+    if config.arrival is not None:
+        arrival = convert_time_to_isodate(config.arrival)
+    if config.departure is not None:
+        departure = convert_time_to_isodate(config.departure)
+
+    return (origin, destination, arrival, departure)
+
+
+def build_hass_attribution(sections: dict) -> str | None:
+    """Build a hass frontend ready string out of the attributions."""
+    relevant_attributions = []
+    for section in sections:
+        if (attributions := section.get("attributions")) is not None:
+            for attribution in attributions:
+                if (href := attribution.get("href")) is not None:
+                    relevant_attributions.append(f"{href}")
+                if (text := attribution.get("text")) is not None:
+                    relevant_attributions.append(text)
+    if len(relevant_attributions) > 0:
+        return ",".join(relevant_attributions)
+    return None
+
+
+def total_transit_distance(sections: dict) -> int:
+    """Calculate the total distance over all sections."""
+    distance = 0
+    for section in sections:
+        distance += section["travelSummary"]["length"]
+    return distance
+
+
+def total_transit_duration(sections: dict) -> int:
+    """Calculate the total duration over all sections."""
+    duration = 0
+    for section in sections:
+        duration += section["travelSummary"]["duration"]
+    return duration
+
+
+def convert_time_to_isodate(simple_time: time) -> str:
+    """Take a time like 08:00:00 and combine it with the current date."""
```

Comment:
You are correct. This is the way it is supposed to work. It's also used in `google_travel_time` and `waze_travel_time`. Im thinking of writing a set of tests for this method and move it to the datetimeutils

Reply:
I could improve the docstring:

```python
def convert_time_to_isodate(simple_time: time) -> str:
    """Take a time like 08:00:00 and return the full datetime iso formatted string representation of the next occurrance.

    If it currently is 07:59:59 or earlier this will return 08:00:00 with todays date.

    If it is currently 08:00:01 or later this will return 08:00:00 with tomorrows date.
    """
    combined = datetime.combine(dt.start_of_local_day(), simple_time)
    if combined < datetime.now():
        combined = combined + timedelta(days=1)
    return combined.isoformat()
```

But this would be explaining what the function does. Kind of an antipattern.

I could also rename the function to something like `iso_date_time_of_next_occurrance`

Reply:
Another idea is to rename to something that explains the logic it is doing e.g. `next_datetime` so it may not need as much elaboration on the doc string? You can make the caller handle `isoformat()`. Anyway, this can happen in a separate PR later if desired.